feat(grandpa): implement neighbor tracker and message handling

jimjbrettj opened the pull request:
## Changes

<!-- Brief list of functional changes -->
This PR adds the neighbor messaging protocol. This includes the ability to handle received neighbor messages, the ability to correctly send neighbor messages, as well as tracking the state of our neighbors (that we use when deciding who to send messages too),

## Tests

<!-- Detail how to run relevant tests to the changes -->

```sh
go test -tags integration github.com/ChainSafe/gossamer
```

## Issues
#2931

<!-- Write the issue number(s), for example: #123 -->

jimjbrettj committed:
wip/neighborMsg handling

jimjbrettj committed:
wip/catch up request flow

jimjbrettj committed:
wip

jimjbrettj committed:
wip/impl neighbor tracking

jimjbrettj committed:
neighbor message tracker

jimjbrettj committed:
remove changes not needed

jimjbrettj committed:
remove minor diff

EclesioMeloJunior commented on lib/grandpa/message_handler.go:
```diff
@@ -82,8 +78,12 @@ func (h *MessageHandler) handleMessage(from peer.ID, m GrandpaMessage) (network.
-func (*MessageHandler) handleNeighbourMessage(_ *NeighbourPacketV1) error {
-	// TODO(#2931)
+func (h *MessageHandler) handleNeighbourMessage(packet *NeighbourPacketV1, from peer.ID) error {
+	logger.Debugf("handleing neighbor message from peer %v with set id %v and round %v", from, packet.SetID, packet.Round)
```
```suggestion
	logger.Debugf("handling neighbor message from peer %s with set id %d and round %d", from.ShortString(), packet.SetID, packet.Round)
```

EclesioMeloJunior commented on lib/grandpa/neighbor_tracker.go:
```diff
@@ -0,0 +1,121 @@
+type neighborState struct {
+	setID uint64
+	round uint64
+	//highestFinalized uint32 not sure if i need this or not
```
Yeah, I would say that this info can be used to check incoming neighbor messages, eg. the neighbor should aways send a highest finalized block greater than the one stored here

EclesioMeloJunior commented on lib/grandpa/neighbor_tracker.go:
```diff
@@ -0,0 +1,121 @@
+
+func (nt *NeighborTracker) Stop() {
+	nt.grandpa.blockState.FreeFinalisedNotifierChannel(nt.finalizationCha)
+	nt.stoppedNeighbor <- struct{}{}
```
```suggestion
	close(nt.stoppedNeighbor)
```

EclesioMeloJunior commented on lib/grandpa/neighbor_tracker.go:
```diff
@@ -0,0 +1,121 @@
+func (nt *NeighborTracker) Stop() {
+	nt.grandpa.blockState.FreeFinalisedNotifierChannel(nt.finalizationCha)
+	nt.stoppedNeighbor <- struct{}{}
+	close(nt.neighborMsgChan)
```
I don't think we should close this channel here as the `*NeighborTracker` is not the writer of this channel, so I would leave the closing to the writer

EclesioMeloJunior commented on lib/grandpa/neighbor_tracker.go:
```diff
@@ -0,0 +1,121 @@
+func (nt *NeighborTracker) run() {
+	logger.Info("starting neighbor tracker")
+	// https://github.com/paritytech/polkadot-sdk/blob/08498f5473351c3d2f8eacbe1bfd7bc6d3a2ef8d/substrate/client/consensus/grandpa/src/communication/mod.rs#L73
+	const duration = time.Minute * 2
```
I would suggest to make this a package lvl constant `NeighborBroadcastPeriod`

EclesioMeloJunior commented on lib/grandpa/neighbor_tracker.go:
```diff
@@ -0,0 +1,121 @@
+func (nt *NeighborTracker) BroadcastNeighborMsg() error {
+	logger.Warnf("braodcasting neighbor message to relevant peers")
+	for id, peerState := range nt.peerview {
+		if !(peerState.round < nt.currentRound || peerState.setID < nt.currentSetID) {
```
```suggestion
		if peerState.round >= nt.currentRound && peerState.setID >= nt.currentSetID {
```

EclesioMeloJunior commented on lib/grandpa/neighbor_tracker.go:
```diff
@@ -0,0 +1,121 @@
+	logger.Warnf("braodcasting neighbor message to relevant peers")
+	for id, peerState := range nt.peerview {
+		if !(peerState.round < nt.currentRound || peerState.setID < nt.currentSetID) {
+			packet := NeighbourPacketV1{
+				Round:  nt.currentRound,
+				SetID:  nt.currentSetID,
+				Number: nt.highestFinalized,
+			}
+
+			cm, err := packet.ToConsensusMessage()
+			if err != nil {
+				return fmt.Errorf("converting NeighbourPacketV1 to network message: %w", err)
+			}
```
can we place the message creation/encoding to out of the loop?

```go
packet := NeighbourPacketV1{
    Round:  nt.currentRound,
    SetID:  nt.currentSetID,
    Number: nt.highestFinalized,
}

cm, err := packet.ToConsensusMessage()
if err != nil {
    return fmt.Errorf("converting NeighbourPacketV1 to network message: %w", err)
}

for ...
```

jimjbrettj replied:
so true thanks for this catch

jimjbrettj committed:
wip/feedback

jimjbrettj committed:
respond to feedback

jimjbrettj committed:
add unit tests

EclesioMeloJunior commented on lib/grandpa/neighbor_tracker.go:
```diff
@@ -0,0 +1,140 @@
+func (nt *NeighborTracker) run() {
+	logger.Info("starting neighbor tracker")
+	const duration = NeighborBroadcastPeriod
+	ticker := time.NewTicker(duration)
```
```suggestion
	ticker := time.NewTicker(NeighborBroadcastPeriod)
```

EclesioMeloJunior reviewed (approved):
LGTM, just a small comment over the usage of a constant.

jimjbrettj committed:
use const for ticker period

haikoschol commented on lib/grandpa/neighbor_tracker.go:
```diff
@@ -0,0 +1,139 @@
+}
+
+func (nt *NeighborTracker) UpdateState(setID uint64, round uint64, highestFinalized uint32) error {
+	if nt == nil {
```
Why check this? Under what conditions does it happen?

jimjbrettj replied:
I had put it there as a precaution, but honestly, I do not know how it would arise. Will remove it

haikoschol commented on lib/grandpa/neighbor_tracker.go:
```diff
@@ -0,0 +1,139 @@
+}
+
+func (nt *NeighborTracker) BroadcastNeighborMsg() error {
+	logger.Warnf("braodcasting neighbor message to relevant peers")
```
Why `warn` instead of `info` or `trace`?

jimjbrettj replied:
Thanks for this catch! I was using `warnf` for my own testing and forgot to change it back

jimjbrettj replied:
This is not a very useful log actually, so will just remove it to not pollute

haikoschol commented on lib/grandpa/neighbor_tracker.go:
```diff
@@ -0,0 +1,139 @@
+		return fmt.Errorf("converting NeighbourPacketV1 to network message: %w", err)
+	}
+	for id, peerState := range nt.peerview {
+		if peerState.round >= nt.currentRound && peerState.setID >= nt.currentSetID {
```
Where does this check come from? I couldn't find something like that in the spec or the Substrate code.

jimjbrettj replied:
[Here](https://spec.polkadot.network/chap-networking#sect-grandpa-neighbor-msg) they say "The sender should take the recipient's state into account and avoid sending messages to peers that are using different voter sets or are in a different round" so this is what I was basing it off of. Let me know if I missed something in my interpretation

haikoschol replied:
I read that too and wondered whether that translates to the `if` condition. But I don't have much context/knowledge about GRANDPA in general so figured I'd ask about it. Anyway, let's keep it like this and revisit if necessary.

haikoschol reviewed (approved):
Just a few nits/questions.

jimjbrettj committed:
feedback and lint

haikoschol commented on lib/grandpa/message_handler.go:
```diff
@@ -78,8 +78,9 @@ func (h *MessageHandler) handleMessage(from peer.ID, m GrandpaMessage) (network.
 
-func (h *MessageHandler) handleNeighbourMessage(packet *NeighbourPacketV1, from peer.ID) error {
-	logger.Debugf("handling neighbor message from peer %v with set id %v and round %v", from.ShortString(), packet.SetID, packet.Round)
+func (h *MessageHandler) handleNeighbourMessage(packet *NeighbourPacketV1, from peer.ID) error { //nolint
```
Is the `//nolint` for something other than the "neighbour" spelling?

jimjbrettj replied:
Yeah it is because error will always be nill. I didn't want to change this to avoid `return nil, nil` in the caller. But now i see this already happens, so i updated to remove returning an error here

jimjbrettj committed:
remove return err from handleNeighbourMSg

jimjbrettj committed:
fix lint and license

timwu20 commented on lib/grandpa/neighbor_tracker.go:
```diff
@@ -0,0 +1,137 @@
+	highestFinalized uint32
+}
+
+type NeighborTracker struct {
```
is there a need to make this public?

jimjbrettj replied:
nope, good catch

timwu20 commented on lib/grandpa/neighbor_tracker.go:
```diff
@@ -0,0 +1,137 @@
+}
+
+func (nt *NeighborTracker) UpdatePeer(p peer.ID, setID uint64, round uint64, highestFinalized uint32) error {
+	if nt.peerview == nil {
+		return fmt.Errorf("neighbour tracker has nil peer tracker")
+	}
```
this is unnecessary code, or should be a panic.  It's understood if you use the constructor to initialize this type, you don't need to add these checks.

timwu20 reviewed (changes requested):
There should be more tests regarding running the `NeighborTracker` and passing in messages via channel. `NeighborTracker.BroadcastNeighborMsg` should be tested.  

I think one of the integration tests are broken as well based on the changes.

jimjbrettj committed:
test for broadcast neighbour msg

jimjbrettj committed:
test to update peer while running

jimjbrettj committed:
fix TestHandleNetworkMessage

jimjbrettj committed:
fix grandpa integration test

jimjbrettj commented:
@timwu20 Responded to feedback, fixed broken integration tests, and added tests for 1) start/stop 2) updating peer state via channel. 

I did not add testing for the finalization channel because no state is actually updated based on it so I didnt know what to assert. Since this channel is used/tested in other parts of the code I thought this is okay, but let me know if/how this should be updated.

jimjbrettj committed:
fix linting

jimjbrettj committed:
fix race conditions

timwu20 commented on lib/grandpa/grandpa.go:
```diff
@@ -151,20 +156,26 @@ func NewService(cfg *Config) (*Service, error) {
 	s.messageHandler = NewMessageHandler(s, s.blockState, cfg.Telemetry)
 	s.tracker = newTracker(s.blockState, s.messageHandler)
 	s.paused.Store(false)
+
```
```suggestion
```

timwu20 commented on lib/grandpa/neighbor_tracker.go:
```diff
@@ -0,0 +1,143 @@
+	"github.com/libp2p/go-libp2p/core/peer"
+)
+
+// https://github.com/paritytech/polkadot-sdk/blob/08498f5473351c3d2f8eacbe1bfd7bc6d3a2ef8d/substrate/client/consensus/grandpa/src/communication/mod.rs#L73 //nolint
```
```suggestion
// How often to rebroadcast neighbor packets, in cases where no new packets are created.
```

timwu20 commented on lib/grandpa/neighbor_tracker.go:
```diff
@@ -0,0 +1,143 @@
+
+func (nt *neighborTracker) Stop() {
+	nt.grandpa.blockState.FreeFinalisedNotifierChannel(nt.finalizationCha)
+	close(nt.stoppedNeighbor)
```
You should wait on `nt.run()` to finish in this `Stop` method.

jimjbrettj replied:
Nice call, will add

timwu20 commented on lib/grandpa/network_integration_test.go:
```diff
@@ -42,6 +42,11 @@ func TestHandleNetworkMessage(t *testing.T) {
 
 	gs, st := newTestService(t, aliceKeyPair)
 
+	//err = gs.Start()
+	//require.NoError(t, err)
```
```suggestion
```

jimjbrettj committed:
respond to most of feedback

jimjbrettj committed:
add wait group to neighbor tracker

timwu20 commented on lib/grandpa/neighbor_tracker.go:
```diff
@@ -0,0 +1,152 @@
+	nt.wg.Add(1)
+	go func() {
+		nt.run()
+		nt.wg.Done()
```
put this as a defer in `run()`

timwu20 commented on lib/grandpa/neighbor_tracker.go:
```diff
@@ -0,0 +1,152 @@
+}
+
+func (nt *neighborTracker) Start() {
+	//go nt.run()
```
```suggestion
```

jimjbrettj committed:
move wg.Done() into a defer
